tools/src/scie_pants/pants_version.py

```diff
@@ -162,5 +182,120 @@ def configure_version():
         pants_config.write_text(tomlkit.dumps(config))
 
     return configure_version, determine_tag_version(
-        ptex, pants_version, find_links_dir, github_api_bearer_token
+        ptex, pants_version, find_links_dir, github_api_bearer_token, bootstrap_urls_path
     )
+
+
+PYTHON_IDS = {
+    # N.B.: These values must match the lift TOML interpreter ids.
+    # Important: all pythons used in pants_python_versions.json must be represented in this list.
+    "cp38": "cpython38",
+    "cp39": "cpython39",
+    "cp310": "cpython310",
+    "cp311": "cpython311",
+}
+
+
+def determine_pex_url_and_python_id(
+    ptex: Ptex,
+    version: Version,
+    bootstrap_urls_path: str | None,
+) -> tuple[str, str]:
+    uname = os.uname()
+    platform = f"{uname.sysname.lower()}_{uname.machine.lower()}"
+    res = get_pex_url_and_python_id(ptex, version, platform, bootstrap_urls_path)
+    if not res:
+        bootstrap_urls_info = (
+            (
+                f"It may also be a missing Pants PEX entry in the PANTS_BOOTSTRAP_URLS file {bootstrap_urls_path} for "
+                f"a name like `pants.{version}-cpNN-{platform}.pex`.\n\n"
+            )
+            if bootstrap_urls_path
+            else ""
+        )
+        fatal(
+            f"Unknown Pants release: {version}.\n\n"
+            "If this is unexpected (you are using a known good Pants version), try upgrading scie-pants first.\n"
+            f"It may also be that the platform {platform} isn't supported for this version of Pants.\n"
+            + bootstrap_urls_info
+            + "To get help, please visit: https://www.pantsbuild.org/community/getting-help"
+        )
+    pex_url, python = res
+    if python not in PYTHON_IDS:
+        # Should not happen... but if we mess up, this is a nicer error message rather than blowing up.
+        fatal(f"This version of scie-pants does not support {python!r}.")
+    return pex_url, PYTHON_IDS[python]
+
+
+def get_pex_url_and_python_id(
+    ptex: Ptex,
+    version: Version,
+    platform: str,
+    bootstrap_urls_path: str | None,
+) -> tuple[str, str] | None:
+    if bootstrap_urls_path:
+        bootstrap_urls = json.loads(Path(bootstrap_urls_path).read_text())
+        ptex_urls = bootstrap_urls.get("ptex")
+        if ptex_urls is None:
+            raise ValueError(
+                f"Missing 'ptex' key in PANTS_BOOTSTRAP_URLS file: {bootstrap_urls_path}"
+            )
+    else:
+        ptex_urls = None
+
+    py = get_python_id_for_pants_version(version)
+    if py:
+        pex_url = get_download_url(version, platform, py, ptex_urls)
+        if pex_url:
+            return pex_url, py
+
+    # Else, try all known Pythons...
+    for maybe_py in PYTHON_IDS.keys():
+        pex_url = get_download_url(version, platform, maybe_py, ptex_urls)
+        if pex_url:
+            return pex_url, maybe_py
```

Comment:
For this fallback, we could take a different tack:

1. hit the GH API `/repos/pantsbuild/pants/releases/releasee_[version}/assets` (https://docs.github.com/en/rest/releases/assets?apiVersion=2022-11-28#list-release-assets) to get a list of all assets
2. If not 200: if 404 => surface as unknown version error to user; if some other problem => surface that error more directly to users (e.g. connection error, 5xx from GitHub)
3. Look through the assets for the one matching the appropriate platform and extract the python identifier from that?

This likely simplifies the error reporting and does only a single request, rather than first trying cp38, then cp39, before eventually getting to the newer versions.

If the API urls have different rate limits to the one we're `HEAD`ing, this is potentially a problem... but I think we already pass around a `github_api_bearer_token` that gives users options, and we can also advise "please upgrade your scie-pants" since the new version is likely to have the URL included.

What do you think?

Reply:
I'm about to dive back in here.. regarding hitting the GH api, that was in the previous iteration but was dropped in favor of just going with we "know" should work to avoid eating on the API rate limit.. for me, I feel either way is OK; I guess we just need to pick our poison here. (or, maybe go with a hybrid solution, perhaps we could do that iteratively though, to make it easier to reason about once we have a working flow in place..?)

Reply:
> rather than first trying cp38, then cp39, before eventually getting to the newer versions.

besides using the API (which I too feel could be the preferable alt), we could also list the pythons in reverse order, so we try with the newest python first, more likely to get a hit sooner.